eclair-core/src/main/scala/fr/acinq/eclair/payment/send/PaymentLifecycle.scala

```diff
@@ -262,43 +262,51 @@ class PaymentLifecycle(nodeParams: NodeParams, cfg: SendPaymentConfig, router: A
+    val extraEdges1 = data.route.hops.find(_.nodeId == nodeId) match {
+      case Some(hop) => hop.params match {
+        case ChannelRelayParams.FromAnnouncement(u) =>
+          if (u.shortChannelId != failure.update.shortChannelId) {
+            // it is possible that nodes in the route prefer using a different channel (to the same N+1 node) than the one we requested, that's fine
+            log.info("received an update for a different channel than the one we asked: requested={} actual={} update={}", u.shortChannelId, failure.update.shortChannelId, failure.update)
+          } else if (Announcements.areSame(u, failure.update)) {
+            // node returned the exact same update we used, this can happen e.g. if the channel is imbalanced
+            // in that case, let's temporarily exclude the channel from future routes, giving it time to recover
+            log.info("received exact same update from nodeId={}, excluding the channel from futures routes", nodeId)
+            router ! ExcludeChannel(ChannelDesc(u.shortChannelId, nodeId, hop.nextNodeId))
+          } else if (PaymentFailure.hasAlreadyFailedOnce(nodeId, data.failures)) {
+            // this node had already given us a new channel update and is still unhappy, it is probably messing with us, let's exclude it
+            log.warning("it is the second time nodeId={} answers with a new update, excluding it: old={} new={}", nodeId, u, failure.update)
+            router ! ExcludeChannel(ChannelDesc(u.shortChannelId, nodeId, hop.nextNodeId))
+          } else {
+            log.info("got a new update for shortChannelId={}: old={} new={}", u.shortChannelId, u, failure.update)
+          }
+          data.c.extraEdges
+        case ChannelRelayParams.FromHint(_) =>
+          log.info("received an update for a routing hint (shortChannelId={} nodeId={} enabled={} update={})", failure.update.shortChannelId, nodeId, failure.update.channelFlags.isEnabled, failure.update)
+          if (failure.update.channelFlags.isEnabled) {
+            data.c.extraEdges.map {
+              case edge: BasicEdge if edge.sourceNodeId == nodeId && edge.targetNodeId == hop.nextNodeId => edge.update(failure.update)
+              case edge: BasicEdge => edge
+            }
+          } else {
+            // if the channel is disabled, we temporarily exclude it: this is necessary because the routing hint doesn't
+            // contain channel flags to indicate that it's disabled
+            // we want the exclusion to be router-wide so that sister payments in the case of MPP are aware the channel is faulty
+            data.c.extraEdges
+              .find { case edge: BasicEdge => edge.sourceNodeId == nodeId && edge.targetNodeId == hop.nextNodeId }
+              .foreach { case edge: BasicEdge => router ! ExcludeChannel(ChannelDesc(edge.shortChannelId, edge.sourceNodeId, edge.targetNodeId)) }
```

Comment:
I was pretty sure that `extraEdges` took precedence over `ignoredEdges` in `dijkstraShortestPath`, but apparently not. The built-in router-wide ban for AMP is pretty cool.